eclipse-collections-api/src/main/java/org/eclipse/collections/api/map/FixedSizeMap.java

```diff
@@ -70,7 +70,18 @@
     @Override
     default FixedSizeMap<K, V> withMap(Map<? extends K, ? extends V> map)
     {
-        this.putAll(map);
-        return this;
```

Comment:
This is a strange inconsistency we have. `withKeyValue` on a fixed-size map is capable of replacing a value at a key without growing the map, but put() always throws.

Reply:
`put()` has to return boolean on a `V` (old value), so it's not possible to have a proper implementation on `FixedSizedMap`. On the other hand, anything that returns a map is entirely implementable. I don't see why we wouldn't implement these...

Reply:
I'm not sure I understand this point. Just to focus on put(), I'm wondering why the implementation in SingletonMap isn't something like:

```java
    @Override
    public V put(K key, V value)
    {
        if (Objects.equals(key, this.key1))
        {
            V result = this.value1;
            this.value1 = value;
            return result;
        }
        throw new UnsupportedOperationException("Cannot call put() on " + this.getClass().getSimpleName());
    }
```

The actual implementation in the superclass is:
```java
    @Override
    public V put(K key, V value)
    {
        throw new UnsupportedOperationException("Cannot call put() on " + this.getClass().getSimpleName());
    }
```

Reply:
I don't see an issue with changing `put` to allow mutation of the value for the same key. (My point was that `put` can't work for the general case, i.e., an arbitrary key). The other methods work just fine for arbitrary parameters, so throwing is particularly strange there.

For some reason, I thought the small maps were immutable, probably remembering really ancient code? On the mutable side, mutations an be allowed as long as the map size doesn't change.

Reply:
Tagging @nsardana-bny @nikhilnanivadekar @donraab for attention.

Reply:
These classes are ancient and not consistent with their mutable or immutable counterparts. The small maps are immutable. The small lists are not. I took a look at `FixedSizeList` and `FixedSizeSet` to see how their `withAll` methods are implemented. I think we should make the `withMapIterable` method consistent and return a new `Map`, except maybe in the case where the map passed in is empty. I'm not keen on the empty handling below, but might be kind of hard to change this behavior if it is depended on anywhere.

FixedSizeSet:

```
public MutableSet<T> withAll(Iterable<? extends T> elements)
{
    if (Iterate.isEmpty(elements))
    {
        return this;
    }
    return Sets.fixedSize.ofAll(this.toList().withAll(elements));
}
```
FixedSizeList:

```
public MutableList<T> withAll(Iterable<? extends T> elements)
{
    if (Iterate.isEmpty(elements))
    {
        return this;
    }
    return Lists.fixedSize.ofAll(this.toList().withAll(elements));
}
```

We should also change the result type for both of these methods (and the other with method) to be `FixedSizeList` or `FixedSizeSet`.

Reply:
> The small maps are immutable.

That's what I vaguely remembered, but I then checked the implementation. Not only does `SingletonMap` extend mutable super classes, but already has a method that mutates its value:

```java
    @Override
    public MutableMap<K, V> withKeyValue(K addKey, V addValue)
    {
        // Map behavior specifies that if you put in a duplicate key, you replace the value
        if (Objects.equals(this.key1, addKey))
        {
            this.value1 = addValue;
            return this;
        }
        return new DoubletonMap<>(this.key1, this.value1, addKey, addValue);
    }
```
(I just realized @motlin pointed out the same thing above).
Given that these are in the mutable side and are already mutable, I would suggest we should make them as consistent with the rest of the mutable implementations as possible.
The mutable map `with` methods mutate the instance, so I have no issue with `withAll` empty checks.